File: solutions/opus-1/opus/python/calc.py
import sys


def tokenize(s):
    toks = []
    i, n = 0, len(s)
    while i < n:
        c = s[i]
        if c.isdigit():
            j = i
            while j < n and s[j].isdigit():
                j += 1
            toks.append(int(s[i:j]))
            i = j
        elif c in "+-*/()":
            toks.append(c)
            i += 1
        else:
            i += 1
    return toks


def tdiv(a, b):
    q = abs(a) // abs(b)
    return q if (a < 0) == (b < 0) else -q


def evaluate(toks):
    pos = 0

    def expr():
        nonlocal pos
        v = term()
        while pos < len(toks) and toks[pos] in ("+", "-"):
            op = toks[pos]
            pos += 1
            r = term()
            v = v + r if op == "+" else v - r
        return v

    def term():
        nonlocal pos
        v = factor()
        while pos < len(toks) and toks[pos] in ("*", "/"):
            op = toks[pos]
            pos += 1
            r = factor()
            v = v * r if op == "*" else tdiv(v, r)
        return v

    def factor():
        nonlocal pos
        t = toks[pos]
        pos += 1
        if t == "(":
            v = expr()
            pos += 1
            return v
        return t

    return expr()


def main():
    sys.setrecursionlimit(1000000)
    out = []
    for line in sys.stdin.read().splitlines():
        toks = tokenize(line)
        if not toks:
            continue
        out.append(str(evaluate(toks)))
    sys.stdout.write("\n".join(out) + ("\n" if out else ""))


main()

File: solutions/opus-1/opus/c/calc.c
#include <stdio.h>
#include <string.h>
#include <ctype.h>

static const char *p;

static void skip(void) { while (*p == ' ' || *p == '\t') p++; }

static long long expr(void);

static long long factor(void) {
    skip();
    if (*p == '(') {
        p++;
        long long v = expr();
        skip();
        if (*p == ')') p++;
        return v;
    }
    long long v = 0;
    while (isdigit((unsigned char)*p)) v = v * 10 + (*p++ - '0');
    return v;
}

static long long term(void) {
    long long v = factor();
    for (;;) {
        skip();
        if (*p == '*') { p++; v *= factor(); }
        else if (*p == '/') { p++; long long d = factor(); v = d ? v / d : 0; }
        else return v;
    }
}

static long long expr(void) {
    long long v = term();
    for (;;) {
        skip();
        if (*p == '+') { p++; v += term(); }
        else if (*p == '-') { p++; v -= term(); }
        else return v;
    }
}

static char line[1 << 20];

int main(void) {
    while (fgets(line, sizeof line, stdin)) {
        size_t n = strlen(line);
        while (n && (line[n-1] == '\n' || line[n-1] == '\r' || line[n-1] == ' ' || line[n-1] == '\t')) line[--n] = 0;
        p = line;
        skip();
        if (!*p) continue;
        printf("%lld\n", expr());
    }
    return 0;
}
